sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1848,6 +1848,54 @@ class Dataset[T] private[sql](
   @scala.annotation.varargs
   def agg(expr: Column, exprs: Column*): DataFrame = groupBy().agg(expr, exprs : _*)
 
+ /**
+  * Define (named) metrics to observe on the Dataset. This method returns an 'observed' Dataset
+  * that returns the same result as the input, with the following guarantees:
+  * - It will compute the defined aggregates (metrics) on all the data that is flowing through the
+  *   Dataset at that point.
+  * - It will report the value of the defined aggregate columns as soon as we reach a completion
+  *   point. A completion point is either the end of a query (batch mode) or the end of a streaming
+  *   epoch. The value of the aggregates only reflects the data processed since the previous
+  *   completion point.
+  * Please note that continuous execution is currently not supported.
+  *
+  * The metrics columns must either contain a literal (e.g. lit(42)), or should contain one or
+  * more aggregate functions (e.g. sum(a) or sum(a + b) + avg(c) - lit(1)). Expressions that
+  * contain references to the input Dataset's columns must always be wrapped in an aggregate
+  * function.
+  *
+  * A user can observe these metrics by either adding
+  * [[org.apache.spark.sql.streaming.StreamingQueryListener]] or a
+  * [[org.apache.spark.sql.util.QueryExecutionListener]] to the spark session.
```

Comment:
Can we have an example with `QueryExecutionListener` too?

Reply:
I'll raise a PR for this, as I found a slight bug in example for streaming as well.

Reply:
Raised #27046. Please note that the patch relies on the current status.

Reply:
So you want to encourage users to use an unstable API?

Reply:
I would like to show the documented usage.

Reply:
and mark this as unstable too .. given that it relies on an unstable APIs.

Reply:
Let's sort it out a bit.

Was the batch use case excluded by intention due to the use of DeveloperApi? If that's the intention, then yes #27046 can just fix the slight bug on streaming example.

If the concern is focused on having QueryExecution as private API and exposing it to DeveloperApi, I think it's time to fix that, as no one would complain for the change on major release

Reply:
We shouldn't document unstable features unless you commit to stabilizing them in the short term. Please don't add that example. A savy developer will know how work with this, and can weight the risk involved themselves.

The same goes for unstable. Please don't mark it as unstable, the function itself is stable. It is stable for streaming. The batch case relies on `QueryExecutionListener` and isn't, however that API is marked as such so there shouldn't be any confusion there.

@HeartSaVioR it would great if we can fix the streaming example.

Reply:
OK thanks for clarification. Totally makes sense. Will update the PR there.

Reply:
I would consider `Dataset.observe` as a whole matrix observing feature, rather than controlling stability for each API as each segment, as this PR logically propose all as single PR. `Dataset.observe` API alone does nothing.

It is `@Evolving` for streaming, which is not stable yet, and it is `@DeveloperApi` and/or `@Unstable` for batch which is unstable. If we go a bit aggressively, we can mark `@Evolving`. If we go a bit conservatively, we should mark `@Unstable` or `@DeveloperApi`.

Reply:
Okay, I offline discussed with @hvanhovell. I am okie.